Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h

```diff
@@ -0,0 +1,873 @@
+//
+// $URL$
+// $Id$
+// 
+//
+// Author(s)     : Andreas Fabri, Fernando Cacciola
+
+#ifndef CGAL_GRAPH_TRAITS_TRIANGULATION_DATA_STRUCTURE_2_H
+#define CGAL_GRAPH_TRAITS_TRIANGULATION_DATA_STRUCTURE_2_H
+
+#include <functional>
+
+// include this to avoid a VC15 warning
+#include <CGAL/boost/graph/named_function_params.h>
+
+#include <boost/config.hpp>
+#include <boost/iterator_adaptors.hpp>
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/properties.hpp>
+
+#include <CGAL/boost/graph/properties.h>
+#include <CGAL/Triangulation_data_structure_2.h>
+#include <CGAL/Iterator_range.h>
+
+// The functions and classes in this file allows the user to
+// treat a CGAL Triangulation_data_structure_2 object as a boost graph "as is". No
+// wrapper is needed for the Triangulation_data_structure_2 object.
+
+
+namespace CGAL {
+
+  namespace detail {
+
+template < class T, class EdgeBase >
+class Edge : public EdgeBase {
+
+public:
+  typedef typename T::Face_handle Face_handle ;
+  
+  Edge()
+  {}
+
+  Edge(Face_handle  fh, int i)
+    : EdgeBase(fh,i)
+  {}
+  
+  Edge(const EdgeBase& e)
+    : EdgeBase(e)
+  {}
+
+  Edge(const Edge& e)
+    : EdgeBase(e)
+  {}
+
+  Edge&
+  operator=(const Edge& e)
+  {
+    this->first = e.first;
+    this->second = e.second;
+    return *this;
+  }
+
+  friend std::size_t hash_value(const Edge& e)
+  {
+    if (e.first==Face_handle()) return 0;
+    return hash_value(e.first<e.first->neighbor(e.second)?
+                      e.first:e.first->neighbor(e.second));
+  }
+
+  bool operator==(const Edge& other) const
+  {
+    if((this->first == other.first)&&(this->second == other.second)) return true;
+    Face_handle fh = this->first->neighbor(this->second);
+    if(other.first != fh) return false;
+    int i = fh->index(this->first);
+    return (other.second == i);
+  }
+
+  bool operator!=(Edge& other) const
+  {
+    return ! (*this == other);
+  }
+};
+
+template <class Circ, class E>
+class Out_edge_circulator : public Circ
+{
+private:
+  mutable E e;
+
+public:
+
+  typedef E value_type;
+  typedef E* pointer;
+  typedef E& reference;
+
+  Out_edge_circulator()
+    : Circ()
+  {}
+
+  Out_edge_circulator(Circ c)
+    : Circ(c)
+  {}
+
+  const E& operator*() const
+  {
+    E ed = static_cast<const Circ*>(this)->operator*();
+    e = E(ed.first->neighbor(ed.second), ed.first->neighbor(ed.second)->index(ed.first));
+    return e;
+  }
+};
+ 
+template <class Circ, class E>
+class In_edge_circulator : public Circ
+{
+private:
+  mutable E e;
+
+public:
+
+  typedef E value_type;
+  typedef E* pointer;
+  typedef E& reference;
+
+  In_edge_circulator()
+    : Circ()
+  {}
+
+  In_edge_circulator(Circ c)
+    : Circ(c)
+  {}
+
+  const E& operator*() const
+{
+    typename Circ::value_type ed = static_cast<const Circ*>(this)->operator*();
+    e = E(ed);
+    return e;
+  }
+};
+  
+
+  //  The vertex iterator of the bgl must evaluate to a vertex handle, not to a vertex
+template < class T>
```

Comment:
I guess `Prevent_deref` could be used all over the place, that is HDS, Arrangement, T2, Polyhedron.
Can we do that in another PR?